Mathlib/LinearAlgebra/RootSystem/Defs.lean

```diff
@@ -280,6 +283,68 @@ lemma isReduced_iff : P.IsReduced ↔ ∀ i j : ι, i ≠ j →
     · exact Or.inl (congrArg P.root h')
     · exact Or.inr (h i j h' hLin)
 
+/-- The `Weyl group` of a root pairing is the group of automorphisms of the weight space generated
+by reflections in roots. For finite root systems, https://en.wikipedia.org/wiki/Weyl_group has
+an exposition. -/
+def weylGroup : Subgroup (M ≃ₗ[R] M) :=
```

Comment:
The definition you suggest would be a substantial improvement, if I could prove that it worked (i.e., that the permutation representation is always faithful).  I'm still thinking about that, so perhaps this PR was a bit premature.

Reply:
I had a few minutes this morning and decided to thought a bit more about `reflection_perm_eq_reflection_perm_iff'`. I believe the following is true:
```
lemma reflection_perm_eq_reflection_perm_iff' [CharZero R] [IsDomain R] [NoZeroSMulDivisors R M]
    (hij : P.pairing j i ≠ 0) :
    P.reflection_perm i = P.reflection_perm j ↔ P.reflection i = P.reflection j := by
```
because if $$r_i$$ and $$r_j$$ are the two reflections and $$n_ij$$ are the pairing constants, then:
$$r_i = r_j \iff \v{alpha_i}\otmes\alpha_i = \v{alpha}_j\otmes\alpha_j \iff n_{ii}n_{ji}\v{alpha}_i\otmes\alpha_i = n_{ii}n_{ji}\v{alpha}_j\otmes\alpha_j $$ and $$n_{ii}n_{ji}\v{alpha}_i\otmes\alpha_i = n_{ii}n_{ji}\v{alpha}_j\otmes\alpha_j = (n_{ii}\v{\alpha}_i) \otimes (n_{ji}\alpha_i) = (n_{ji}\v{\alpha}_j) \otimes (n_{jj}\alpha_j)$$
because the ith and jth permutations agree.

It also seems to me that we can prove the result for finite reduced root pairings. Sorry that's rather rushed, I probably won't have time to think about this again for a few days (by which time you may have had a chance to return to it and sort it all out).

Unfortunately I have to start work now.

Reply:
That's great.  So far, I've been able to show that `Injective fun (x : M) => (2 : R) • x` implies the equivalence.

Reply:
I had a few minutes over lunch and managed to prove:
```
lemma reflection_perm_eq_reflection_perm_iff [CharZero R] [IsDomain R] [NoZeroSMulDivisors R M] :
    P.reflection_perm i = P.reflection_perm j ↔ P.reflection i = P.reflection j := by
  refine ⟨fun h ↦ ?_, fun h ↦ Equiv.ext fun k ↦ P.root.injective <| by simp [h]⟩
  have hr : P.pairing j i • P.root i = P.pairing j j • P.root j := by
    have : P.reflection i (P.root j) = P.reflection j (P.root j) := by
      simp only [← root_reflection_perm, h]
    simpa only [reflection, Module.reflection_apply, sub_right_inj]
  have hc : P.pairing i i • P.coroot i = P.pairing j i • P.coroot j := by
    have : (P.coreflection i) (P.coroot i) = (P.coreflection j) (P.coroot i) := by
      simp only [← coroot_reflection_perm, h]
    simpa only [coreflection, Module.reflection_apply, sub_right_inj]
  have hij : 2 * P.pairing j i ≠ 0 := by
    suffices P.pairing j i ≠ 0 by simpa
    exact fun contra ↦ P.ne_zero j <| by simpa [contra, eq_comm (a := (0 : M))] using hr
  suffices (2 * P.pairing j i) • (P.flip.toLin (P.coroot i)).smulRight (P.root i) =
           (2 * P.pairing j i) • (P.flip.toLin (P.coroot j)).smulRight (P.root j) by
    simpa [reflection, Module.reflection, Module.preReflection] using
      smul_right_injective (M := M →ₗ[R] M) hij this
  suffices (P.flip.toLin (P.pairing i i • P.coroot i)).smulRight (P.pairing j i • P.root i) =
           (P.flip.toLin (P.pairing j i • P.coroot j)).smulRight (P.pairing j j • P.root j) by
    ext x
    replace this := LinearMap.congr_fun this x
    simp only [pairing_same, map_smul, ← mul_smul, LinearMap.coe_smulRight, LinearMap.smul_apply,
      smul_eq_mul] at this ⊢
    convert this using 2 <;> ring
  rw [hr, hc]

lemma _root_.RootSystem.reflection_perm_eq_reflection_perm_iff (P : RootSystem ι R M N) (i j : ι) :
    P.reflection_perm i = P.reflection_perm j ↔ P.reflection i = P.reflection j := by
  refine ⟨fun h ↦ ?_, fun h ↦ Equiv.ext fun k ↦ P.root.injective <| by simp [h]⟩
  ext x
  have hx : x ∈ span R (range P.root) := by simp [P.span_eq_top]
  induction hx using Submodule.span_induction' with
  | mem x hx =>
    obtain ⟨k, rfl⟩ := hx
    simp only [← root_reflection_perm, h]
  | zero => simp
  | add x _ y _ hx hy => simp [hx, hy]
  | smul t x _ hx => simp [hx]
```
The fact that there doesn't seem to be a perfect unification of assumptions makes me wonder if we should add an axiom to `RootPairing` (though it's premature to do so now I think).

Reply:
> That's great. So far, I've been able to show that `Injective fun (x : M) => (2 : R) • x` implies the equivalence.

Oh this is great. Note:
```lean
example [CharZero R] [NoZeroSMulDivisors R M] :
    Injective fun (x : M) => (2 : R) • x :=
  smul_right_injective _ two_ne_zero
```

Reply:
Specifically, I have a proof (not too far from your argument) that 
`P.reflection_perm i = P.reflection_perm j → ∀ (x : M), 2 • (P.reflection i x -  P.reflection j x) = 0`
Also, I found a counterexample in char 2. Let `i j` be the only roots, linearly independent, with reflection acting trivially. If we extend `M` suitably, we can take some `x` on which `P.reflection i x = x` and `P.reflection j x ≠ x`.

Reply:
In light of the non-faithfulness of the permutation representation in general, I think it is best to keep the current definition of `weylGroup`.